design/dpop.md

```diff
@@ -0,0 +1,308 @@
+
+_([6.2. JWK Thumbprint Confirmation Method in Token Introspection][8])_
+
+Classes/methods affected:
+
+* org.keycloak.protocol.oidc.AccessTokenIntrospectionProvider
+  * introspect()
+
+#### Metadata Endpoint
+
+With DPoP enabled, a new parameter, `dpop_signing_alg_values_supported`, must be included into the authorization server metadata.
+
+_([5.1. Authorization Server Metadata][10])_
+
+Classes/methods affected:
+
+* org.keycloak.protocol.oidc.OIDCWellKnownProvider
+	* getConfig()
+* org.keycloak.protocol.oidc.representations.OIDCConfigurationRepresentation
+
+#### UserInfo Endpoint
+
+The spec doesn’t directly mention the UserInfo endpoint, but it clearly falls into the category of bearer token protected resources:
+
+> The UserInfo Endpoint is an OAuth 2.0 Protected Resource that returns Claims about the authenticated End-User. To obtain the requested Claims about the End-User, the Client makes a request to the UserInfo Endpoint using an Access Token obtained through OpenID Connect Authentication.
+
+_(OpenID Connect Core 1.0,_ [_5.3. UserInfo Endpoint_][11]_)_
+
+Therefore, UserInfo endpoint must also be made DPoP-aware.
+
+Classes/methods affected:
+
+* org.keycloak.protocol.oidc.endpoints.UserInfoEndpoint
+	* issueUserInfo()
+
+#### Logout Endpoint
+
+The `POST` variant of logout invocation uses refresh token, which could be DPoP-bound. Therefore, Logout endpoint must also be made DPoP-aware.
+
+Classes/methods affected:
+
+* org.keycloak.protocol.oidc.endpoints.LogoutEndpoint
+	* logoutToken()
+
+#### Other Endpoints
+
+The following endpoints in Keycloak are public:
+
+* Server Discovery
+* Server JWK set
+* Authorization
+
+The following endpoints don’t use bearer tokens, but rather client credentials:
+
+* Token Introspection
+* Token Revocation
+
+The following endpoints use cookie authentication:
+
+* Check Session iframe
+
+Therefore, no other endpoints should be affected.
+
+#### Client Registration
+Client registration should support DPoP-related parameters inside client metadata.
+
+Classes/methods affected:
+
+* org.keycloak.services.clientregistration.oidc.DescriptionConverter
+* org.keycloak.representations.oidc.OIDCClientRepresentation
+
+### Models
+
+#### Access Token
+
+New property, `jkt`, should be recognized as a member of the `cnf` claim.
+
+Classes/methods affected:
+
+* org.keycloak.representations.AccessToken.CertConf
+	* rename to Conf or Confirmation; introduce `keyThumbprint` property
+
+#### Client
+
+There are the following DPoP-related client settings:
+
+* DPoP mode (enabled/optional/disabled)
+* DPoP proof lifetime (integer)
+* DPoP clock skew (integer)
+
+Those could be backed by client attributes, hence no changes to the client model and DB schema.
+
+Classes/methods affected:
+
+* org.keycloak.protocol.oidc.OIDCAdvancedConfigWrapper
+
+### Adapters
```

Comment:
TBH, I would leave adapters as the lowest priority unless there is an explicit request for adding DPoP support to them. Do you have a need for adapters support for your deployment?

Reply:
Agreed; no, we do not have any requirements for the adapters. Will reflect that in the doc.